commit ff3ca3743a00af749258cc242457b648d65a1537
Author: Stefan Liebler <stli@linux.ibm.com>
Date:   Tue Dec 18 13:57:10 2018 +0100

    S390: Refactor strlen ifunc handling.
    
    The ifunc handling for strlen is adjusted in order to omit ifunc
    variants if those will never be used as the minimum architecture level
    already supports newer CPUs by default.
    Glibc internal calls will then also use the "newer" ifunc variant.
    
    ChangeLog:
    
            * sysdeps/s390/multiarch/Makefile
            (sysdep_routines): Remove strlen variants.
            * sysdeps/s390/Makefile (sysdep_routines): Add strlen variants.
            * sysdeps/s390/multiarch/ifunc-impl-list.c
            (__libc_ifunc_impl_list): Refactor ifunc handling for strlen.
            * sysdeps/s390/multiarch/strlen-c.c: Move to ...
            * sysdeps/s390/strlen-c.c: ... here and adjust ifunc handling.
            * sysdeps/s390/multiarch/strlen-vx.S: Move to ...
            * sysdeps/s390/strlen-vx.S: ... here and adjust ifunc handling.
            * sysdeps/s390/multiarch/strlen.c: Move to ...
            * sysdeps/s390/strlen.c: ... here and adjust ifunc handling.
            * sysdeps/s390/ifunc-strlen.h: New file.

diff --git a/sysdeps/s390/Makefile b/sysdeps/s390/Makefile
index 47d606d3d5d99274..600d8e629df7090e 100644
--- a/sysdeps/s390/Makefile
+++ b/sysdeps/s390/Makefile
@@ -36,5 +36,6 @@ sysdep_routines += bzero memset memset-z900 \
 		   mempcpy memcpy memcpy-z900 \
 		   memmove memmove-c \
 		   strstr strstr-vx strstr-c \
-		   memmem memmem-vx memmem-c
+		   memmem memmem-vx memmem-c \
+		   strlen strlen-vx strlen-c
 endif
diff --git a/sysdeps/s390/ifunc-strlen.h b/sysdeps/s390/ifunc-strlen.h
new file mode 100644
index 0000000000000000..f2070596636f29a9
--- /dev/null
+++ b/sysdeps/s390/ifunc-strlen.h
@@ -0,0 +1,52 @@
+/* strlen variant information on S/390 version.
+   Copyright (C) 2018 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#if defined USE_MULTIARCH && IS_IN (libc)		\
+  && ! defined HAVE_S390_MIN_Z13_ZARCH_ASM_SUPPORT
+# define HAVE_STRLEN_IFUNC	1
+#else
+# define HAVE_STRLEN_IFUNC	0
+#endif
+
+#ifdef HAVE_S390_VX_ASM_SUPPORT
+# define HAVE_STRLEN_IFUNC_AND_VX_SUPPORT HAVE_STRLEN_IFUNC
+#else
+# define HAVE_STRLEN_IFUNC_AND_VX_SUPPORT 0
+#endif
+
+#if defined HAVE_S390_MIN_Z13_ZARCH_ASM_SUPPORT
+# define STRLEN_DEFAULT		STRLEN_Z13
+# define HAVE_STRLEN_C		0
+# define HAVE_STRLEN_Z13	1
+#else
+# define STRLEN_DEFAULT		STRLEN_C
+# define HAVE_STRLEN_C		1
+# define HAVE_STRLEN_Z13	HAVE_STRLEN_IFUNC_AND_VX_SUPPORT
+#endif
+
+#if HAVE_STRLEN_C
+# define STRLEN_C		__strlen_c
+#else
+# define STRLEN_C		NULL
+#endif
+
+#if HAVE_STRLEN_Z13
+# define STRLEN_Z13		__strlen_vx
+#else
+# define STRLEN_Z13		NULL
+#endif
diff --git a/sysdeps/s390/multiarch/Makefile b/sysdeps/s390/multiarch/Makefile
index 24949cd3a88b8015..601523919c235f76 100644
--- a/sysdeps/s390/multiarch/Makefile
+++ b/sysdeps/s390/multiarch/Makefile
@@ -1,6 +1,5 @@
 ifeq ($(subdir),string)
-sysdep_routines += strlen strlen-vx strlen-c \
-		   strnlen strnlen-vx strnlen-c \
+sysdep_routines += strnlen strnlen-vx strnlen-c \
 		   strcpy strcpy-vx \
 		   stpcpy stpcpy-vx stpcpy-c \
 		   strncpy strncpy-vx \
diff --git a/sysdeps/s390/multiarch/ifunc-impl-list.c b/sysdeps/s390/multiarch/ifunc-impl-list.c
index da8696d917abf51c..c531be4bc7eb3f55 100644
--- a/sysdeps/s390/multiarch/ifunc-impl-list.c
+++ b/sysdeps/s390/multiarch/ifunc-impl-list.c
@@ -26,6 +26,7 @@
 #include <ifunc-memcpy.h>
 #include <ifunc-strstr.h>
 #include <ifunc-memmem.h>
+#include <ifunc-strlen.h>
 
 /* Maximum number of IFUNC implementations.  */
 #define MAX_IFUNC	3
@@ -164,6 +165,18 @@ __libc_ifunc_impl_list (const char *name, struct libc_ifunc_impl *array,
 		)
 #endif /* HAVE_MEMMEM_IFUNC  */
 
+#if HAVE_STRLEN_IFUNC
+    IFUNC_IMPL (i, name, strlen,
+# if HAVE_STRLEN_Z13
+		IFUNC_IMPL_ADD (array, i, strlen,
+				dl_hwcap & HWCAP_S390_VX, STRLEN_Z13)
+# endif
+# if HAVE_STRLEN_C
+		IFUNC_IMPL_ADD (array, i, strlen, 1, STRLEN_C)
+# endif
+		)
+#endif /* HAVE_STRLEN_IFUNC  */
+
 #ifdef HAVE_S390_VX_ASM_SUPPORT
 
 # define IFUNC_VX_IMPL(FUNC)						\
@@ -172,7 +185,6 @@ __libc_ifunc_impl_list (const char *name, struct libc_ifunc_impl *array,
 			      __##FUNC##_vx)				\
 	      IFUNC_IMPL_ADD (array, i, FUNC, 1, __##FUNC##_c))
 
-  IFUNC_VX_IMPL (strlen);
   IFUNC_VX_IMPL (wcslen);
 
   IFUNC_VX_IMPL (strnlen);
diff --git a/sysdeps/s390/multiarch/strlen-c.c b/sysdeps/s390/strlen-c.c
similarity index 78%
rename from sysdeps/s390/multiarch/strlen-c.c
rename to sysdeps/s390/strlen-c.c
index a2c8e43624a9bc91..b4569701af96f4a9 100644
--- a/sysdeps/s390/multiarch/strlen-c.c
+++ b/sysdeps/s390/strlen-c.c
@@ -16,13 +16,17 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)
-# define STRLEN  __strlen_c
-# ifdef SHARED
-#  undef libc_hidden_builtin_def
-#  define libc_hidden_builtin_def(name)			\
+#include <ifunc-strlen.h>
+
+#if HAVE_STRLEN_C
+# if HAVE_STRLEN_IFUNC
+#  define STRLEN STRLEN_C
+#  if defined SHARED && IS_IN (libc)
+#   undef libc_hidden_builtin_def
+#   define libc_hidden_builtin_def(name)		\
   __hidden_ver1 (__strlen_c, __GI_strlen, __strlen_c);
-# endif /* SHARED */
+#  endif
+# endif
 
 # include <string/strlen.c>
-#endif /* HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc) */
+#endif
diff --git a/sysdeps/s390/multiarch/strlen-vx.S b/sysdeps/s390/strlen-vx.S
similarity index 90%
rename from sysdeps/s390/multiarch/strlen-vx.S
rename to sysdeps/s390/strlen-vx.S
index 9308b332371dcdaa..39ef43107d11ec73 100644
--- a/sysdeps/s390/multiarch/strlen-vx.S
+++ b/sysdeps/s390/strlen-vx.S
@@ -16,7 +16,8 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)
+#include <ifunc-strlen.h>
+#if HAVE_STRLEN_Z13
 
 # include "sysdep.h"
 # include "asm-syntax.h"
@@ -34,7 +35,7 @@
    -r5=current_len and return_value
    -v16=part of s
 */
-ENTRY(__strlen_vx)
+ENTRY(STRLEN_Z13)
 	.machine "z13"
 	.machinemode "zarch_nohighgprs"
 
@@ -80,5 +81,13 @@ ENTRY(__strlen_vx)
 	vlgvb	%r2,%v16,7	/* Load byte index of zero.  */
 	algr	%r2,%r5
 	br	%r14
-END(__strlen_vx)
-#endif /* HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc) */
+END(STRLEN_Z13)
+
+# if ! HAVE_STRLEN_IFUNC
+strong_alias (STRLEN_Z13, strlen)
+# endif
+
+# if ! HAVE_STRLEN_C && defined SHARED && IS_IN (libc)
+strong_alias (STRLEN_Z13, __GI_strlen)
+# endif
+#endif
diff --git a/sysdeps/s390/multiarch/strlen.c b/sysdeps/s390/strlen.c
similarity index 69%
rename from sysdeps/s390/multiarch/strlen.c
rename to sysdeps/s390/strlen.c
index 0edf8b7d0231cf31..6ba0fe86fe9789f0 100644
--- a/sysdeps/s390/multiarch/strlen.c
+++ b/sysdeps/s390/strlen.c
@@ -16,14 +16,25 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)
+#include <ifunc-strlen.h>
+
+#if HAVE_STRLEN_IFUNC
 # define strlen __redirect_strlen
 # include <string.h>
 # include <ifunc-resolve.h>
 # undef strlen
 
-s390_vx_libc_ifunc2_redirected (__redirect_strlen, __strlen, strlen)
+# if HAVE_STRLEN_C
+extern __typeof (__redirect_strlen) STRLEN_C attribute_hidden;
+# endif
+
+# if HAVE_STRLEN_Z13
+extern __typeof (__redirect_strlen) STRLEN_Z13 attribute_hidden;
+# endif
 
-#else
-# include <string/strlen.c>
-#endif /* !(defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)) */
+s390_libc_ifunc_expr (__redirect_strlen, strlen,
+		      (HAVE_STRLEN_Z13 && (hwcap & HWCAP_S390_VX))
+		      ? STRLEN_Z13
+		      : STRLEN_DEFAULT
+		      )
+#endif
